File: swimming-contest-software/Server/src/main/java/server/repository/dbRepository/RaceDBRepository.java
package server.repository.dbRepository;

import model.domain.entities.Race;
import model.domain.enums.SwimmingDistances;
import model.domain.enums.SwimmingStyles;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import server.repository.interfaces.RaceRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RaceDBRepository implements RaceRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger =  LogManager.getLogger();

    public RaceDBRepository(Properties properties) {
        logger.info("Initialising RaceDBRepository with properties {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Race findRaceByDistanceAndStyle(SwimmingDistances swimmingDistance, SwimmingStyles swimmingStyle) {
        logger.traceEntry("findRaceByDistanceAndStyle(swimmingDistance = {}, swimmingStyle = {})",
                swimmingDistance,
                swimmingStyle);

        Race race = null;

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select id, swimmersNumber from main.Races where distance = ? and style = ?;"
        )) {
            preparedStatement.setInt(1, SwimmingDistances.integerFromDistance(swimmingDistance));
            preparedStatement.setInt(2, SwimmingStyles.integerFromStyle(swimmingStyle));
            ResultSet resultSet = preparedStatement.executeQuery();
            resultSet.next();
            Integer id = resultSet.getInt("id");
            Integer swimmersNumber = resultSet.getInt("swimmersNumber");
            race = new Race(id, swimmingDistance, swimmingStyle, swimmersNumber);
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return logger.traceExit("Result: race = {}", race);
    }

    @Override
    public List<Race> findAllRaces() {
        logger.traceEntry("findAllRaces()");

        List<Race> races = new ArrayList<>();

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select * from main.Races;"
        )) {
            ResultSet resultSet = preparedStatement.executeQuery();
            while(resultSet.next()) {
                Integer id = resultSet.getInt("id");
                Integer swimmingDistance = resultSet.getInt("distance");
                Integer swimmingStyle = resultSet.getInt("style");
                Integer swimmersNumber = resultSet.getInt("swimmersNumber");
                Race race = new Race(id,
                        SwimmingDistances.distanceFromInteger(swimmingDistance),
                        SwimmingStyles.styleFromInteger(swimmingStyle),
                        swimmersNumber);
                races.add(race);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return logger.traceExit("Result: allRaces  = {}", races);
    }

    @Override
    public Integer add(Race elem) {
        return null;
    }

    @Override
    public void delete(Race elem) {

    }

    @Override
    public void update(Race elem, Integer id) {

    }

    @Override
    public Race findById(Integer id) {
        logger.traceEntry("findById(id = {})", id);
        Race race = null;

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select distance, style, swimmersNumber from main.Races where id = ?;"
        )) {
            preparedStatement.setInt(1, id);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                Integer distance = resultSet.getInt("distance");
                Integer style = resultSet.getInt("style");
                Integer swimmersNo = resultSet.getInt("swimmersNumber");
                race = new Race(id, SwimmingDistances.distanceFromInteger(distance), SwimmingStyles.styleFromInteger(style), swimmersNo);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return logger.traceExit("Result: race = {}", race);
    }
}

File: Server/Repository/DBRepository/RaceDBRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using Model.Domain.Entities;
using Model.Domain.Enums;
using Server.Utils;

namespace Server.Repository.DBRepository;

public class RaceDBRepository : IRaceRepository
{
    private static readonly ILog Logger = LogManager.GetLogger("RaceDbRepository");
    private readonly IDictionary<string, string> properties;

    public RaceDBRepository(IDictionary<string, string> properties)
    {
        Logger.InfoFormat("Initialising RaceDBRepository...");
        this.properties = properties;
    }

    public int Add(Race elem)
    {
        throw new NotImplementedException();
    }

    public void Delete(Race elem)
    {
        throw new NotImplementedException();
    }

    public void Update(Race elem, int id)
    {
        throw new NotImplementedException();
    }

    public Race FindById(int id)
    {
        Logger.InfoFormat("FindById(id = {0})", id);
        Race race = null;

        var connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select distance, style, swimmersNumber from Races where id = @id;";

            var paramID = command.CreateParameter();
            paramID.ParameterName = "id";
            paramID.Value = id;
            command.Parameters.Add(paramID);


            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    var swimmingDistance =
                        SwimmingDistancesMethods.DistanceFromInteger(dataReader.GetInt32(0));
                    var swimmingStyle = SwimmingStylesMethods.StyleFromInteger(dataReader.GetInt32(1));
                    var swimmersNumber = dataReader.GetInt32(2);
                    race = new Race(id, swimmingDistance, swimmingStyle, swimmersNumber);
                }
            }
        }

        Logger.InfoFormat("Result: race = {0}", race);

        return race;
    }

    public Race FindRaceByDistanceAndStyle(SwimmingDistance swimmingDistance, SwimmingStyle swimmingStyle)
    {
        Logger.InfoFormat("FindRaceByDistanceAndStyle(distance = {0}, style = {1})", swimmingDistance, swimmingStyle);
        Race race = null;

        var connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select id, swimmersNumber from Races where distance = @distance and style = @style;";

            var paramDistance = command.CreateParameter();
            paramDistance.ParameterName = "distance";
            paramDistance.Value = SwimmingDistancesMethods.IntegerFromDistance(swimmingDistance);
            command.Parameters.Add(paramDistance);

            var paramStyle = command.CreateParameter();
            paramStyle.ParameterName = "style";
            paramStyle.Value = SwimmingStylesMethods.IntegerFromStyle(swimmingStyle);
            command.Parameters.Add(paramStyle);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    var id = dataReader.GetInt32(0);
                    var swimmersNumber = dataReader.GetInt32(1);
                    race = new Race(id, swimmingDistance, swimmingStyle, swimmersNumber);
                }
            }
        }

        Logger.InfoFormat("Result: race = {0}", race);
        return race;
    }

    public List<Race> FindAllRaces()
    {
        Logger.InfoFormat("FindAllRaces()");
        var races = new List<Race>();

        var connection = DbUtils.GetConnection(properties);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select id, distance, style, swimmersNumber from Races;";

            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    var id = dataReader.GetInt32(0);
                    var distance = SwimmingDistancesMethods.DistanceFromInteger(dataReader.GetInt32(1));
                    var style = SwimmingStylesMethods.StyleFromInteger(dataReader.GetInt32(2));
                    var swimmersNumber = dataReader.GetInt32(3);
                    races.Add(new Race(id, distance, style, swimmersNumber));
                }
            }
        }

        Logger.InfoFormat("Result: races = {0}", races);
        return races;
    }
}